cli/tests/test_fix_command.rs

```diff
@@ -0,0 +1,513 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::os::unix::fs::PermissionsExt;
+use std::path::PathBuf;
+
+use jj_lib::file_util::try_symlink;
+
+use crate::common::TestEnvironment;
+
+/// Set up a repo where the `jj fix` command uses the fake editor with the given
+/// flags.
+fn init_with_fake_formatter(args: &str) -> (TestEnvironment, PathBuf) {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+    let formatter_path = assert_cmd::cargo::cargo_bin("fake-formatter");
+    assert!(formatter_path.is_file());
+    let escaped_formatter_path = formatter_path.to_str().unwrap().replace('\\', r"\\");
+    test_env.add_config(&format!(
+        r#"fix.tool-command = "{} {}""#,
+        &escaped_formatter_path, &args
+    ));
+    (test_env, repo_path)
+}
+
+#[test]
+fn test_fix_no_config() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["fix", "-s", "@"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Config error: configuration property "fix.tool-command" not found
+    For help, see https://github.com/martinvonz/jj/blob/main/docs/config.md.
+    "###);
+}
+
+#[test]
+fn test_fix_empty_commit() {
+    let (test_env, repo_path) = init_with_fake_formatter("--reverse");
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["fix", "-s", "@"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Fixed 0 commits of 1 checked.
+    Nothing changed.
+    "###);
+}
+
+#[test]
+fn test_fix_leaf_commit() {
+    let (test_env, repo_path) = init_with_fake_formatter("--reverse");
```

Comment:
`--reverse` is a quite weird formatter since it's not idempotent. I think it makes sense to test that in some cases, but should we use a more well-behaved formatter in both of these tests? How about making `fake-formatter` uppercase text by default and have the `--reverse` argument only for testing odd behavior? Uppercasing also makes it easier to notice correspondence between input and output.

Btw, if you want more tests of weird formatters, I suppose one formatter behavior could be to add a prefix to every line. That would impact even empty files. If you made the prefix e.g. "formatted: ", it would also make it easy to notice correspondence.

Reply:
These are all good points. Uppercase is actually one of the things we used in hg. I think we do want to focus on supporting idempotent tools, even if we have coverage for cyclic or diverging behavior. Reversing is a bit of a risk in tests that expect an odd number of changes. I think it makes sense to encode in the file content information about how many times the tool was executed (like "add a prefix").

I initially reached for `rev`/`tac` because I was worried about portability, but that's irrelevant now.